`kfp-v2` bundle tests fail during recurring run

### Bug Description

Intermittently with #601, tests fail during test_create_and_monitor_recurring_run:

```
AssertionError: assert 'FAILED' == 'SUCCEEDED'
  
  - SUCCEEDED
  + FAILED
```

Looking at the juju status, it looks like node could be potentially out of space

```
Model     Controller                Cloud/Region        Version  SLA          Timestamp
kubeflow  github-pr-38fa2-microk8s  microk8s/localhost  3.4.6    unsupported  14:34:1[7](https://github.com/canonical/kfp-operators/actions/runs/11838249242/job/32986966903?pr=583#step:7:8)Z

App                      Version                  Status   Scale  Charm                    Channel       Rev  Address         Exposed  Message
argo-controller                                   active       1  argo-controller          latest/edge   596  10.152.1[8](https://github.com/canonical/kfp-operators/actions/runs/11838249242/job/32986966903?pr=583#step:7:9)3.140  no       
envoy                                             active       1  envoy                    latest/edge   307  10.152.183.161  no       
istio-ingressgateway                              active       1  istio-gateway            latest/edge  1287  10.152.183.151  no       
istio-pilot                                       active       1  istio-pilot              latest/edge  1235  10.152.183.45   no       
kfp-api                                           active       1  kfp-api                                  0  10.152.183.241  no       
kfp-db                   8.0.37-0ubuntu0.22.04.3  active       1  mysql-k8s                8.0/stable    180  10.152.183.8[9](https://github.com/canonical/kfp-operators/actions/runs/11838249242/job/32986966903?pr=583#step:7:10)   no       
kfp-metadata-writer                               active       1  kfp-metadata-writer                      0  [10](https://github.com/canonical/kfp-operators/actions/runs/11838249242/job/32986966903?pr=583#step:7:11).152.183.114  no       
kfp-persistence                                   active       1  kfp-persistence                          0  10.152.183.247  no       
kfp-profile-controller                            active       1  kfp-profile-controller                   0  10.152.183.139  no       
kfp-schedwf                                       active       1  kfp-schedwf                              0  10.152.183.214  no       
kfp-ui                                            active       1  kfp-ui                                   0  10.152.183.248  no       
kfp-viewer                                        active       1  kfp-viewer                               0  10.152.183.170  no       
kfp-viz                                           active       1  kfp-viz                                  0  10.152.183.20   no       
kubeflow-profiles                                 active       1  kubeflow-profiles        latest/edge   456  10.152.183.58   no       
kubeflow-roles                                    active       1  kubeflow-roles           latest/edge   264  10.152.183.57   no       
metacontroller-operator                           active       1  metacontroller-operator  latest/edge   349  10.152.183.185  no       
minio                    res:oci-image@220b31a    blocked    0/1  minio                    latest/edge   376  10.152.183.226  no       0/1 nodes are available: 1 node(s) had untolerated taint {node.kubernetes.io/disk-pressure: }. preemption: 0/1 nodes ...
mlmd                                              active       1  mlmd                     latest/edge   243  10.152.183.225  no       

Unit                        Workload  Agent  Address      Ports          Message
argo-controller/0*          active    idle   10.1.2[11](https://github.com/canonical/kfp-operators/actions/runs/11838249242/job/32986966903?pr=583#step:7:12).74                 
envoy/0*                    active    idle   10.1.211.75                 
istio-ingressgateway/0*     active    idle   10.1.211.76                 
istio-pilot/0*              active    idle   10.1.211.77                 
kfp-api/0*                  active    idle   10.1.211.78                 
kfp-db/0*                   active    idle   10.1.211.82                 Primary
kfp-metadata-writer/0*      active    idle   10.1.211.81                 
kfp-persistence/0*          active    idle   10.1.211.84                 
kfp-profile-controller/0*   active    idle   10.1.211.85                 
kfp-schedwf/0*              active    idle   10.1.[21](https://github.com/canonical/kfp-operators/actions/runs/11838249242/job/32986966903?pr=583#step:7:22)1.86                 
kfp-ui/0*                   active    idle   10.1.211.88                 
kfp-viewer/0*               active    idle   10.1.211.89                 
kfp-viz/0*                  active    idle   10.1.211.90                 
kubeflow-profiles/0*        active    idle   10.1.211.93                 
kubeflow-roles/0*           active    idle   10.1.211.91                 
metacontroller-operator/0*  active    idle   10.1.211.92                 
minio/0*                    unknown   lost   10.1.211.97  9000-9001/TCP  agent lost, see 'juju show-status-log minio/0'
mlmd/0*                     active    idle   10.1.211.95
```

### To Reproduce

Rerun CI from PR https://github.com/canonical/kfp-operators/pull/583

### Environment

Juju 3.4.6
Microk8s 1.29

### Relevant Log Output

```shell
____________________ test_create_and_monitor_recurring_run _____________________
Traceback (most recent call last):
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/runner.py", line 341, in from_call
    result: TResult | None = func()
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/runner.py", line 242, in <lambda>
    lambda: runtest_hook(item=item, **kwds), when=when, reraise=reraise
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 139, in _multicall
    raise exception.with_traceback(exception.__traceback__)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/threadexception.py", line 92, in pytest_runtest_call
    yield from thread_exception_runtest_hook()
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/threadexception.py", line 68, in thread_exception_runtest_hook
    yield
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/unraisableexception.py", line 95, in pytest_runtest_call
    yield from unraisable_exception_runtest_hook()
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/unraisableexception.py", line 70, in unraisable_exception_runtest_hook
    yield
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/logging.py", line 848, in pytest_runtest_call
    yield from self._runtest_for(item, "call")
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/logging.py", line 831, in _runtest_for
    yield
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/capture.py", line 879, in pytest_runtest_call
    return (yield)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/skipping.py", line 257, in pytest_runtest_call
    return (yield)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/runner.py", line 174, in pytest_runtest_call
    item.runtest()
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/python.py", line 1627, in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 182, in _multicall
    return outcome.get_result()
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_result.py", line 100, in get_result
    raise exc.with_traceback(exc.__traceback__)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/python.py", line 159, in pytest_pyfunc_call
    result = testfunction(**testargs)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pytest_asyncio/plugin.py", line 529, in inner
    _loop.run_until_complete(task)
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/home/runner/work/kfp-operators/kfp-operators/tests/integration/test_kfp_functional_v2.py", line 204, in test_create_and_monitor_recurring_run
    assert monitor_response.state == "SUCCEEDED"
AssertionError: assert 'FAILED' == 'SUCCEEDED'
  
  - SUCCEEDED
  + FAILED
```


### Additional Context

_No response_

--- a/.github/workflows/integrate.yaml
+++ b/.github/workflows/integrate.yaml
@@ -169,7 +169,7 @@ jobs:
         run: |
           # Requires the model to be called kubeflow due to kfp-viewer
           juju add-model kubeflow
-          sg snap_microk8s -c "tox -e bundle-integration-${{ matrix.sdk }} -- --model kubeflow --bundle=./tests/integration/bundles/kfp_latest_edge.yaml.j2"
+          sg snap_microk8s -c "tox -e bundle-integration-${{ matrix.sdk }} -- --model kubeflow --bundle=./tests/integration/bundles/kfp_latest_edge.yaml.j2" --charmcraft-clean
 
       - name: Get all
         run: kubectl get all -A

--- a/requirements-integration-v1.in
+++ b/requirements-integration-v1.in
@@ -11,3 +11,4 @@ pytest
 pytest-operator
 pyyaml
 tenacity
+sh

--- a/requirements-integration-v1.txt
+++ b/requirements-integration-v1.txt
@@ -272,6 +272,8 @@ rich==13.7.1
     # via typer
 rsa==4.9
     # via google-auth
+sh==2.1.0
+    # via -r requirements-integration-v1.in
 shellingham==1.5.4
     # via typer
 six==1.16.0

--- a/tests/README.md
+++ b/tests/README.md
@@ -65,5 +65,4 @@ Where,
 * `--model` tells the testing framework which model to deploy charms to
 * `--bundle` is the path to a bundle template that's going to be used during the test execution
 * `--no-build` tells the test suite whether to build charms and run tests against them, or use charms in Charmhub
-* `--no-build` tells the test suite whether to build charms and run tests against them, or use charms in Charmhub
 * `--charmcraft-clean` tells the test suite whether to run `charmcraft clean` and delete the lxc instances after building the charms. If `--no-build` is passed, this has no effect.

--- a/tests/README.md
+++ b/tests/README.md
@@ -17,6 +17,7 @@ This directory has the following structure:
     ├── conftest.py
     ├── helpers
     │   ├── bundle_mgmt.py
+    │   ├── charmcraft.py
     │   ├── k8s_resources.py
     │   └── localize_bundle.py
     ├── kfp_globals.py
@@ -57,10 +58,12 @@ Communication with the KFP API happens using the KFP Python SDK. A `kfp.client` 
 2. Run integration tests against the preferred bundle definition in `integration/bundles`
 
 ```
-tox -e bundle-integration -- --model kubeflow --bundle=./tests/integration/bundles/<bundle_template> <--no-build>
+tox -e bundle-integration -- --model kubeflow --bundle=./tests/integration/bundles/<bundle_template> <--no-build> <--charmcraft-clean>
 ```
 
 Where,
 * `--model` tells the testing framework which model to deploy charms to
 * `--bundle` is the path to a bundle template that's going to be used during the test execution
 * `--no-build` tells the test suite whether to build charms and run tests against them, or use charms in Charmhub
+* `--no-build` tells the test suite whether to build charms and run tests against them, or use charms in Charmhub
+* `--charmcraft-clean` tells the test suite whether to run `charmcraft clean` and delete the lxc instances after building the charms. If `--no-build` is passed, this has no effect.

--- a/tests/integration/conftest.py
+++ b/tests/integration/conftest.py
@@ -108,7 +108,6 @@ def pytest_addoption(parser: Parser):
     parser.addoption(
         "--charmcraft-clean",
         action="store_true",
-        default=False,
         help="Whether to run charmcraft clean and delete lxc instances created by charmcraft."
         "It defaults to False."
     )

--- a/tests/integration/conftest.py
+++ b/tests/integration/conftest.py
@@ -105,3 +105,10 @@ def pytest_addoption(parser: Parser):
         "to render the bundle definition template."
         "If set to False, the integration tests will be run against charms in Charmhub.",
     )
+    parser.addoption(
+        "--charmcraft-clean",
+        action="store_true",
+        default=False,
+        help="Whether to run charmcraft clean and delete lxc instances created by charmcraft."
+        "It defaults to False."
+    )

--- a/tests/integration/helpers/charmcraft.py
+++ b/tests/integration/helpers/charmcraft.py
@@ -1,0 +1,10 @@
+import sh
+
+def charmcraft_clean(charms_paths: list[str]) -> None:
+    """
+    Run `charmcraft clean` for passed paths to charms.
+    """
+    for charm_path in charms_paths:
+        sh.cd(charm_path)
+        sh.charmcraft.clean()
+        sh.cd("-")

--- a/tests/integration/helpers/charmcraft.py
+++ b/tests/integration/helpers/charmcraft.py
@@ -1,10 +1,23 @@
+import logging
+from os import chdir
+from pathlib import Path
+from typing import Dict
+
+from kfp_globals import basedir
+
 import sh
 
-def charmcraft_clean(charms_paths: list[str]) -> None:
+log = logging.getLogger(__name__)
+
+
+def charmcraft_clean(charms_paths: Dict[str, Path]) -> None:
     """
     Run `charmcraft clean` for passed paths to charms.
     """
-    for charm_path in charms_paths:
-        sh.cd(charm_path)
+    pwd = sh.pwd()
+    for charm,charm_path in charms_paths.items():
+        log.info(f"Charmcraft clean {charm}")
+        chdir(charm_path)
         sh.charmcraft.clean()
-        sh.cd("-")
+    # Return to original directory so it doesn't affect tests execution
+    chdir(pwd)

--- a/tests/integration/test_kfp_functional_v1.py
+++ b/tests/integration/test_kfp_functional_v1.py
@@ -9,7 +9,7 @@ from pathlib import Path
 from helpers.bundle_mgmt import render_bundle, deploy_bundle
 from helpers.k8s_resources import apply_manifests, fetch_response
 from helpers.localize_bundle import get_resources_from_charm_file
-from tests.integration.helpers.charmcraft import charmcraft_clean
+from helpers.charmcraft import charmcraft_clean
 from kfp_globals import (
     CHARM_PATH_TEMPLATE,
     KFP_CHARMS,

--- a/tests/integration/test_kfp_functional_v1.py
+++ b/tests/integration/test_kfp_functional_v1.py
@@ -9,6 +9,7 @@ from pathlib import Path
 from helpers.bundle_mgmt import render_bundle, deploy_bundle
 from helpers.k8s_resources import apply_manifests, fetch_response
 from helpers.localize_bundle import get_resources_from_charm_file
+from tests.integration.helpers.charmcraft import charmcraft_clean
 from kfp_globals import (
     CHARM_PATH_TEMPLATE,
     KFP_CHARMS,
@@ -18,6 +19,7 @@ from kfp_globals import (
     SAMPLE_VIEWER,
 )
 
+import sh
 import kfp
 import lightkube
 import pytest
@@ -68,6 +70,7 @@ def create_and_clean_experiment_v1(kfp_client: kfp.Client):
 async def test_build_and_deploy(ops_test: OpsTest, request, lightkube_client):
     """Build and deploy kfp-operators charms."""
     no_build = request.config.getoption("no_build")
+    charmcraft_clean_flag = True if request.config.getoption("--charmcraft-clean") else False
 
     # Immediately raise an error if the model name is not kubeflow
     if ops_test.model_name != "kubeflow":
@@ -92,6 +95,9 @@ async def test_build_and_deploy(ops_test: OpsTest, request, lightkube_client):
             charm_resources = get_resources_from_charm_file(charm_file)
             context.update([(f"{charm.replace('-', '_')}_resources", charm_resources)])
             context.update([(f"{charm.replace('-', '_')}", charm_file)])
+
+        if charmcraft_clean_flag == True:
+            charmcraft_clean(charms_to_build)
 
     # Render kfp-operators bundle file with locally built charms and their resources
     rendered_bundle = render_bundle(

--- a/tests/integration/test_kfp_functional_v2.py
+++ b/tests/integration/test_kfp_functional_v2.py
@@ -83,14 +83,14 @@ async def test_build_and_deploy(ops_test: OpsTest, request, lightkube_client):
     basedir = Path("./").absolute()
 
     # Build the charms we need to build only if --no-build is not set
-    # context = {}
+    context = {}
     if not no_build:
         charms_to_build = {
             charm: Path(CHARM_PATH_TEMPLATE.format(basedir=str(basedir), charm=charm))
             for charm in KFP_CHARMS
         }
         log.info(f"Building charms for: {charms_to_build}")
-        # built_charms = await ops_test.build_charms(*charms_to_build.values())
+        built_charms = await ops_test.build_charms(*charms_to_build.values())
         log.info(f"Built charms: {built_charms}")
 
         for charm, charm_file in built_charms.items():

--- a/tests/integration/test_kfp_functional_v2.py
+++ b/tests/integration/test_kfp_functional_v2.py
@@ -9,7 +9,7 @@ from pathlib import Path
 from helpers.bundle_mgmt import render_bundle, deploy_bundle
 from helpers.k8s_resources import apply_manifests, fetch_response
 from helpers.localize_bundle import get_resources_from_charm_file
-from tests.integration.helpers.charmcraft import charmcraft_clean
+from helpers.charmcraft import charmcraft_clean
 from kfp_globals import (
     CHARM_PATH_TEMPLATE,
     KFP_CHARMS,
@@ -83,14 +83,14 @@ async def test_build_and_deploy(ops_test: OpsTest, request, lightkube_client):
     basedir = Path("./").absolute()
 
     # Build the charms we need to build only if --no-build is not set
-    context = {}
+    # context = {}
     if not no_build:
         charms_to_build = {
             charm: Path(CHARM_PATH_TEMPLATE.format(basedir=str(basedir), charm=charm))
             for charm in KFP_CHARMS
         }
         log.info(f"Building charms for: {charms_to_build}")
-        built_charms = await ops_test.build_charms(*charms_to_build.values())
+        # built_charms = await ops_test.build_charms(*charms_to_build.values())
         log.info(f"Built charms: {built_charms}")
 
         for charm, charm_file in built_charms.items():

--- a/tests/integration/test_kfp_functional_v2.py
+++ b/tests/integration/test_kfp_functional_v2.py
@@ -9,6 +9,7 @@ from pathlib import Path
 from helpers.bundle_mgmt import render_bundle, deploy_bundle
 from helpers.k8s_resources import apply_manifests, fetch_response
 from helpers.localize_bundle import get_resources_from_charm_file
+from tests.integration.helpers.charmcraft import charmcraft_clean
 from kfp_globals import (
     CHARM_PATH_TEMPLATE,
     KFP_CHARMS,
@@ -71,6 +72,7 @@ def create_and_clean_experiment_v2(kfp_client: kfp.Client):
 async def test_build_and_deploy(ops_test: OpsTest, request, lightkube_client):
     """Build and deploy kfp-operators charms."""
     no_build = request.config.getoption("no_build")
+    charmcraft_clean_flag = True if request.config.getoption("--charmcraft-clean") else False
 
     # Immediately raise an error if the model name is not kubeflow
     if ops_test.model_name != "kubeflow":
@@ -95,6 +97,9 @@ async def test_build_and_deploy(ops_test: OpsTest, request, lightkube_client):
             charm_resources = get_resources_from_charm_file(charm_file)
             context.update([(f"{charm.replace('-', '_')}_resources", charm_resources)])
             context.update([(f"{charm.replace('-', '_')}", charm_file)])
+
+        if charmcraft_clean_flag == True:
+            charmcraft_clean(charms_to_build)
 
     # Render kfp-operators bundle file with locally built charms and their resources
     rendered_bundle = render_bundle(

